core/src/main/java/io/grpc/internal/BinaryLog.java

```diff
@@ -0,0 +1,208 @@
+
+  @VisibleForTesting
+  BinaryLog(int maxHeaderBytes, int maxMessageBytes) {
+    this.maxHeaderBytes = maxHeaderBytes;
+    this.maxMessageBytes = maxMessageBytes;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (!(o instanceof BinaryLog)) {
+      return false;
+    }
+    BinaryLog that = (BinaryLog) o;
+    return this.maxHeaderBytes == that.maxHeaderBytes
+        && this.maxMessageBytes == that.maxMessageBytes;
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(maxHeaderBytes, maxMessageBytes);
+  }
+
+  @Override
+  public String toString() {
+    return getClass().getSimpleName() + '['
+        + "maxHeaderBytes=" + maxHeaderBytes + ", "
+        + "maxMessageBytes=" + maxMessageBytes + "]";
+  }
+
+  private static final Factory DEFAULT_FACTORY =
+      new Factory(System.getenv("GRPC_BINARY_LOG_CONFIG"));
+
+  /**
+   * Accepts the fullMethodName and returns the binary log that should be used. The log may be
+   * a log that does nothing.
+   */
+  public static BinaryLog getLog(String fullMethodName) {
+    return DEFAULT_FACTORY.getLog(fullMethodName);
+  }
+
+  static final class Factory {
+    // '*' for global, 'service/*' for service glob, or 'service/method' for fully qualified.
+    private static final Pattern logPatternRe = Pattern.compile("[^{]+");
+    // A curly brace wrapped expression. Will be further matched with the more specified REs below.
+    private static final Pattern logOptionsRe = Pattern.compile("\\{[^}]+}");
+    private static final Pattern configRe = Pattern.compile(
+        String.format("^(%s)(%s)?$", logPatternRe.pattern(), logOptionsRe.pattern()));
+    // Regexes to extract per-binlog options
+    // The form: {m:256}
+    private static final Pattern msgRe = Pattern.compile("\\{m(?::(\\d+))?}");
+    // The form: {h:256}
+    private static final Pattern headerRe = Pattern.compile("\\{h(?::(\\d+))?}");
+    // The form: {h:256,m:256}
+    private static final Pattern bothRe = Pattern.compile("\\{h(?::(\\d+))?;m(?::(\\d+))?}");
+
+    private final boolean enabled;
+    private final Map<String, BinaryLog> perMethodLogs = new HashMap<String, BinaryLog>();
+    private final Map<String, BinaryLog> perServiceLogs = new HashMap<String, BinaryLog>();
+    private final BinaryLog globalLog;
+
+    /**
+     * Accepts a string in the format specified by the binary log spec.
+     */
+    @VisibleForTesting
+    Factory(String configurationString) {
+      if (configurationString == null || configurationString.length() == 0) {
+        globalLog = NOOP_LOG;
+        enabled = false;
+        return;
+      }
+      String[] configurations = configurationString.split(",");
+      BinaryLog globalLog = null;
+      for (String configuration : configurations) {
+        Matcher configMatcher = configRe.matcher(configuration);
+        if (!configMatcher.matches()) {
+          throw new IllegalArgumentException("Bad input: " + configuration);
```

Comment:
Exceptions in this constructor that is initialized statically seem a bad idea. It seems it should be logged and the entry ignored.

Reply:
Now ignoring duplicate entries